File: Question 0109 - Convert Sorted List to Binary Search Tree/src/Solution.java
class Solution {

    private int calcSize(ListNode head)
    {
        ListNode temp = head;
        int count = 0;
        if(head == null) return count;

        while(temp != null)
        {
            temp = temp.next;
            count++;
        }

        return count;
    }

    private int getAtPos(ListNode head, int index)
    {
        ListNode temp = head;
        if(head == null) return -1;

        while (temp != null) {
            if (index == 0)
                return temp.val;

            temp = temp.next;
            index--;
        }

        return -1;
    }
    private TreeNode converter(ListNode head, int start, int end)
    {
        if(start > end) return null;

        int mid = (start + end) / 2;
        TreeNode root = new TreeNode(getAtPos(head, mid));
        root.left = converter(head, start, mid-1);
        root.right = converter(head,mid+1, end);

        return root;
    }
    public TreeNode sortedListToBST(ListNode head) {
        int size = calcSize(head);
        if(head == null) return null;
        else return converter(head, 0, size-1);
    }
}

File: Question 0109 - Convert Sorted List to Binary Search Tree/src/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
private:
    TreeNode *solve(ListNode *head, ListNode *tail)
    {
        if (head == tail)
            return NULL;

        ListNode *slow = head;
        ListNode *fast = head;

        while (fast != tail && fast->next != tail)
        {
            slow = slow->next;
            fast = fast->next->next;
        }

        TreeNode *root = new TreeNode(slow->val);
        root->left = solve(head, slow);
        root->right = solve(slow->next, tail);
        return root;
    }

public:
    TreeNode *sortedListToBST(ListNode *head)
    {
        if (!head)
            return NULL;
        return solve(head, NULL);
    }
};